### PR TITLE
Added Netronome packetgen script for bursty traffic generation.

### PR DESCRIPTION
**Pull request by RobertClucas**
I have added an example script for generating a specific bursty traffic pattern. The pattern is explained in the script, and is suitable specifically for Netronome NFPs. It's able to generate line-rate packet rates, and has numerous configuration options for configuring the traffic pattern.

### CODE REVIEW

File: examples/packetgen/packetgen.lua [hunk 1]
```diff
@@ -0,0 +1,1262 @@
...
+    local macString = string.format("%u:%u:%u:%u:%u:%u",
+      macArray[1], macArray[2], macArray[3],
+      macArray[4], macArray[5], macArray[6]
+    )
+    return macString
+end
+
+-- @brief Generates a random number between 1 - 120 for the IMIX.
+function imixSize() 
+    math.randomseed(os.time())
```
Comment (emmericp) [inline examples/packetgen/packetgen.lua]: reseeding the PRNG for each call is kind of pointless -- even wrong in this case as os.time() has a granularity of 1 second

**Comment by RobertClucas** [comment]
Sorry, I though that I has already pushed the new script changes to the repo. I've added the updates to the script -- these updates have been quite thoroughly tested by our QA team. They are also going to be doing further tests, after which I will add any additional changes.

**Comment by emmericp** [comment]
As I've mentioned via Email: there are a few issues that I'd like to see addressed before merging this; one is the name, "packetgen" sounds very generic.

Can you change this to something less generic? However, I don't have a good idea.

File: examples/packetgen/packetgen.lua [hunk 2]
```diff
@@ -0,0 +1,1454 @@
...
+  )
+  
+  print(statsString)
+end
+
+-- Utilities ------------------------------------------------------------------
+
+-- @brief Converts a number to a hex string representation.
+-- @param num The number to convert to hex.
+function num2hex(num)
```
Comment (emmericp) [inline examples/packetgen/packetgen.lua]: string.format("%x", num)

File: examples/packetgen/packetgen.lua [hunk 3]
```diff
@@ -0,0 +1,1454 @@
...
+        num       = math.floor(num / 16)
+    end
+    if s == '' then s = '0' end
+    return s
+end
+
+-- Converts a mac address in number format (a long) to a string based mac
+-- address of the form: a:b:c:d:e:f
+-- @param mac The mac address to convert
+function convertMacNumberToString(mac)
```
Comment (emmericp) [inline examples/packetgen/packetgen.lua]: can this be moved to utils.lua?

File: examples/packetgen/packetgen.lua [hunk 4]
```diff
@@ -0,0 +1,1454 @@
...
+    end
+    local macString = string.format("%u:%u:%u:%u:%u:%u",
+      macArray[1], macArray[2], macArray[3],
+      macArray[4], macArray[5], macArray[6]
+    )
+    return macString
+end
+
+-- @brief Generates a random number in range [1, 120] for the IMIX.
+function imixSize()
```
Comment (emmericp) [inline examples/packetgen/packetgen.lua]: maybe move to utils.lua as well?

File: examples/packetgen/packetgen.lua [hunk 5]
```diff
@@ -0,0 +1,1454 @@
...
+    elseif rnum <= 50 then     -- 4 in 12 : [11 - 50]  : 40 / 120
+        return 574
+    else                       -- 7 in 12 : [51 - 120] : 70 / 120
+        return 68     
+    end
+end
+
+-- Deep copies a table, and returns the copy.
+-- @param orig The original table to copy.
+function deepCopy(orig)
```
Comment (emmericp) [inline examples/packetgen/packetgen.lua]: move to utils.lua

**Comment by emmericp** [comment]
The timestamping is currently done in Lua which is suboptimal due to garbage collection and jit pauses.
This might not be a problem for your use case but the documentation should mention this.

**Comment by RobertClucas** [comment]
Regarding the name, it's supposed to be generic because it can generate a range of different traffic patterns. I will try to come up with another option though and get back to you.

We have reviewed the script and have decided that we would like to clean the interface and the code itself, to be easier to understand and use. It will be done by the end of the week. When it is, I will add the all  changes you suggested and commit them. I will let you know when that's done.

File: examples/packetgen/packetgen.lua (hunk 3 above)
Comment (RobertClucas) [inline examples/packetgen/packetgen.lua]: Got rid of this in the updates, would you still like me to add it to utils.lua?

File: examples/packetgen/packetgen.lua (hunk 4 above)
Comment (RobertClucas) [inline examples/packetgen/packetgen.lua]: Got rid of this in the updates, would you still like me to add it to utils.lua?

**Comment by RobertClucas** [comment]
Hi Paul,

I have made numerous changes to our packet generator. Please see the pull
request and let me know if there are any further changes you would like me
to make. We have not come up with a better name. We are happy to change it
if you have any suggestions.

Rob

File: examples/packetgen/packetgen.lua (hunk 4 above)
Comment (emmericp) [inline examples/packetgen/packetgen.lua]: no, that's fine

**Comment by emmericp** [comment]
Hi,

looks better. Two minor things:

* can you just change the name of the folder to "netronome-packetgen" or some other prefix. Just to avoid confusion for new users
* you are ignoring ctrl+c and SIGTERM (which is not that bad since moongen just kills the threads anyways if it receives two signals). But this should be easy to fix:
just add an additional check for ``moongen.running()`` in the loops that currently check the state variable, i.e. ``while moongen.running() and state == "canrun" do``

Thanks!

**Comment by RobertClucas** [comment]
Hi Paul,

No problem, I will make those changes. I was actually stuck on that for a
while, will that change mean that one ctrl+c will kill all threads? That's
what we would like, but have not been sure how to do it.

Thanks for the help!

**Comment by RobertClucas** [comment]
Hi Paul,

Just one other thing. Sometimes when running MoonGen we have got the
following message as output:

EAL: Ask a virtual area of ...
EAL: Virtual area found at ...

But many of those messages, say upward of 32. Do you know what may cause
that?

Thanks for any help!

**Comment by RobertClucas** [comment]
Here is some more of the output detailing the error:

```
EAL: Requesting 512 pages of size 2MB from socket 0
EAL: Requesting 512 pages of size 2MB from socket 1
EAL: Can only reserve 459 pages from 1024 requested
Current CONFIG_RTE_MAX_MEMSEG=256 is not enough
Please either increase it or request less amount of memory.
PANIC in rte_eal_init():
```
This only happens on some systems, and not on others. Do you know what may cause this?

**Comment by emmericp** [comment]
This happens due to memory fragmentation, some systems seem more prone to this than others (I've found some enterprisey suse linux to be quite bad with this but never had problems with my Debian and Ubuntu machines).

Rebooting helps.
You can also allocate fewer huge pages (cf. setup-hugetlbfs.sh)

(Note that this doesn't have anything to do with MoonGen but is a DPDK issue)



As for the SIGTERM/INT issue: yes, that should fix ctrl+c.

**Comment by RobertClucas** [comment]
It's only happened on occasion, so as you said, it definitely a system problem.

I've added the other changes you mentioned,  is there anything else?

**Comment by RobertClucas** [comment]
Another question.

MoonGen generates a list of available devices from the PCI HW addresses specified in the dpdk-conf.lua file.

Using:
```
pciwhite={"0000:03:08.0", "0000:03:08.1"}
```
produces:
```
[INFO]  Found 3 usable devices:
   Device 0: AE:91:80:81:FC:09 (Netronome NFP6000 Virtual Function)
   Device 1: 5E:55:00:1B:32:7E (Netronome NFP6000 Virtual Function)
```
and if I swap the order of the PCI HW addresses in the dpdk-conf.lua file, device 0 and device 1 still have the same MAC addresses.

What is the relationship between the PCI HW addresses and MoonGen devices? Is the lowest PCI HW address always assigned to device 0, second lowest to device 1 etc?

**Comment by emmericp** [comment]
Yes, they are ordered by PCI ID

**Pull request merged**